apollo-federation/src/query_plan/conditions.rs

```diff
@@ -32,99 +50,140 @@ pub(crate) enum Conditions {
             };
-            let value = directive.specified_argument_by_name("if").ok_or_else(|| {
-                FederationError::internal(format!(
-                    "missing if argument on @{}",
-                    if negated { "skip" } else { "include" },
-                ))
-            })?;
-            match &**value {
-                Value::Boolean(false) if !negated => return Ok(Self::Boolean(false)),
-                Value::Boolean(true) if negated => return Ok(Self::Boolean(false)),
+
+            match value.as_ref() {
+                // Constant @skip(if: true) can never match
+                Value::Boolean(true) => return Ok(Self::Boolean(false)),
+                // Constant @skip(if: false) always matches
                 Value::Boolean(_) => {}
-                Value::Variable(name) => match variables.entry(name.clone()) {
-                    Entry::Occupied(entry) => {
-                        let previous_negated = *entry.get();
-                        if previous_negated != negated {
-                            return Ok(Self::Boolean(false));
-                        }
-                    }
-                    Entry::Vacant(entry) => {
-                        entry.insert(negated);
+                Value::Variable(name) => {
+                    variables.insert(name.clone(), ConditionKind::Skip);
+                }
+                _ => {
+                    internal_error!("expected boolean or variable `if` argument, got {value}");
+                }
+            }
+        }
+
+        if let Some(include) = directives.get("include") {
+            let Some(value) = include.specified_argument_by_name("if") else {
+                internal_error!("missing @include(if:) argument");
+            };
+
+            match value.as_ref() {
+                // Constant @include(if: false) can never match
+                Value::Boolean(false) => return Ok(Self::Boolean(false)),
+                // Constant @include(if: true) always matches
+                Value::Boolean(true) => {}
+                // If both @skip(if: $var) and @include(if: $var) exist, the condition can also
+                // never match
+                Value::Variable(name) => {
+                    if variables.insert(name.clone(), ConditionKind::Include)
+                        == Some(ConditionKind::Skip)
+                    {
+                        return Ok(Self::Boolean(false));
                     }
-                },
+                }
                 _ => {
-                    return Err(FederationError::internal(format!(
-                        "expected boolean or variable `if` argument, got {value}",
-                    )))
+                    internal_error!("expected boolean or variable `if` argument, got {value}");
                 }
             }
         }
+
         Ok(Self::from_variables(variables))
     }
 
+    // TODO(@goto-bus-stop): what exactly is the difference between this and `Self::merge`?
```

Comment:
Indeed, `merge` makes more sense. It's unclear why `update_with` is so loose in the `Boolean(false)` cases.